src/python/pants/engine/collection.py

```diff
@@ -42,7 +42,7 @@ def __getitem__(self, index: int | slice) -> T | Collection[T]:  # noqa: F811
     def __eq__(self, other: Any) -> bool:
         if self is other:
             return True
-        return type(self) == type(other) and super().__eq__(other)
+        return isinstance(self, type(other)) and super().__eq__(other)
```

Comment:
This is not semantically equivalent. A subclass instance will pass `isinstance` but not the direct type comparison.